### ISSUE TITLE
ContentLength with Body length 0 errors

### ISSUE THREAD

**Issue by tj** [apex/up#276]
The common issue so far seems to be apps which set content-length and don't respond with a body, or partial body etc. The errors from the reverse proxy don't make it clear what's happened.

**Comment by cloudkite** [comment]
Firstly thanks for the awesome tool ❤️ 

Is this the same as `ERRO network error: http: ContentLength=1058 with Body length 0`? getting this error alot would love to be able to dig a bit better and find out the offending url etc.

**Comment by tj** [comment]
@cloudkite thanks! We were thinking that case is a bug in the proxy, I'll try and reproduce it today. If you have any isolated code where it fails reasonably consistently or notice any patterns let me know! 

Do you see any "retrying idempotent request" in the logs?

**Comment by cloudkite** [comment]
@tj dont see any "retrying idempotent request" in the logs

This is what I get in the logs

```
{"fields":{"app":"HelpHero","error":"EOF","plugin":"relay","region":"us-west-2","version":"13"},"level":"error","timestamp":"2017-09-27T05:15:56.263949574Z","message":"network"}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","version":"13"},"level":"warn","timestamp":"2017-09-27T05:15:56.263992087Z","message":"restarting"}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","version":"13"},"level":"info","timestamp":"2017-09-27T05:15:56.264059763Z","message":"found free port 40065"}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","version":"13"},"level":"info","timestamp":"2017-09-27T05:15:56.264087603Z","message":"executing \"cd dist \u0026\u0026 node-v8.4.0-linux-x64/bin/node lib/functions\""}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","version":"13"},"level":"info","timestamp":"2017-09-27T05:15:56.266981961Z","message":"proxy (pid=29) started"}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","version":"13"},"level":"info","timestamp":"2017-09-27T05:15:56.267024325Z","message":"waiting for http://127.0.0.1:40065 to listen (timeout 15s)"}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","restarts":1,"version":"13"},"level":"warn","timestamp":"2017-09-27T05:15:57.593718135Z","message":"restarted"}
{"fields":{"app":"HelpHero","error":"http: ContentLength=1058 with Body length 0","plugin":"relay","region":"us-west-2","version":"13"},"level":"error","timestamp":"2017-09-27T05:15:57.594089195Z","message":"network"}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","version":"13"},"level":"warn","timestamp":"2017-09-27T05:15:57.594116478Z","message":"restarting"}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","version":"13"},"level":"info","timestamp":"2017-09-27T05:15:57.594177518Z","message":"found free port 45525"}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","version":"13"},"level":"info","timestamp":"2017-09-27T05:15:57.594210115Z","message":"executing \"cd dist \u0026\u0026 node-v8.4.0-linux-x64/bin/node lib/functions\""}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","version":"13"},"level":"info","timestamp":"2017-09-27T05:15:57.604005752Z","message":"proxy (pid=36) started"}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","version":"13"},"level":"info","timestamp":"2017-09-27T05:15:57.604043442Z","message":"waiting for http://127.0.0.1:45525 to listen (timeout 15s)"}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","restarts":2,"version":"13"},"level":"warn","timestamp":"2017-09-27T05:15:58.930635714Z","message":"restarted"}
{"fields":{"app":"HelpHero","error":"http: ContentLength=1058 with Body length 0","plugin":"relay","region":"us-west-2","version":"13"},"level":"error","timestamp":"2017-09-27T05:15:58.930997464Z","message":"network"}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","version":"13"},"level":"warn","timestamp":"2017-09-27T05:15:58.931024558Z","message":"restarting"}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","version":"13"},"level":"info","timestamp":"2017-09-27T05:15:58.931085031Z","message":"found free port 38441"}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","version":"13"},"level":"info","timestamp":"2017-09-27T05:15:58.931120843Z","message":"executing \"cd dist \u0026\u0026 node-v8.4.0-linux-x64/bin/node lib/functions\""}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","version":"13"},"level":"info","timestamp":"2017-09-27T05:15:58.931534264Z","message":"proxy (pid=43) started"}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","version":"13"},"level":"info","timestamp":"2017-09-27T05:15:58.931569487Z","message":"waiting for http://127.0.0.1:38441 to listen (timeout 15s)"}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","restarts":3,"version":"13"},"level":"warn","timestamp":"2017-09-27T05:16:00.253728035Z","message":"restarted"}
{"fields":{"app":"HelpHero","duration":3992,"id":"f0dfd800-a342-11e7-967b-c317544628a5","method":"POST","path":"/api/query","plugin":"logs","query":"trace=nav-startVisit","region":"us-west-2","size":0,"stage":"production","status":502,"version":"13"},"level":"error","timestamp":"2017-09-27T05:16:00.254142004Z","message":"response"}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","version":"13"},"level":"warn","timestamp":"2017-09-27T05:16:20.168008382Z","message":"proxy (pid=17) sending SIGKILL"}
```

on some occassion it leads with a  `connection reset by peer` instead of `EOF` 

```
{"fields":{"app":"HelpHero","error":"read tcp 127.0.0.1:48090-\u003e127.0.0.1:39613: read: connection reset by peer","plugin":"relay","region":"us-west-2","version":"13"},"level":"warn","timestamp":"2017-09-27T12:11:24.726339448Z","message":"temporary"}
{"fields":{"app":"HelpHero","error":"http: ContentLength=1058 with Body length 0","plugin":"relay","region":"us-west-2","version":"13"},"level":"error","timestamp":"2017-09-27T12:11:24.826867736Z","message":"network"}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","version":"13"},"level":"warn","timestamp":"2017-09-27T12:11:24.826921335Z","message":"restarting"}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","version":"13"},"level":"info","timestamp":"2017-09-27T12:11:24.826982667Z","message":"found free port 45465"}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","version":"13"},"level":"info","timestamp":"2017-09-27T12:11:24.827018379Z","message":"executing \"cd dist \u0026\u0026 node-v8.4.0-linux-x64/bin/node lib/functions\""}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","version":"13"},"level":"info","timestamp":"2017-09-27T12:11:24.844072155Z","message":"proxy (pid=54) started"}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","version":"13"},"level":"info","timestamp":"2017-09-27T12:11:24.84412175Z","message":"waiting for http://127.0.0.1:45465 to listen (timeout 15s)"}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","restarts":4,"version":"13"},"level":"warn","timestamp":"2017-09-27T12:11:26.173694539Z","message":"restarted"}
{"fields":{"app":"HelpHero","error":"http: ContentLength=1058 with Body length 0","plugin":"relay","region":"us-west-2","version":"13"},"level":"error","timestamp":"2017-09-27T12:11:26.174047008Z","message":"network"}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","version":"13"},"level":"warn","timestamp":"2017-09-27T12:11:26.174073827Z","message":"restarting"}
{"fields":{"name":"HelpHero","type":"server"},"level":"info","timestamp":"2017-09-27T12:13:05.018284557Z","message":"starting"}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","version":"13"},"level":"info","timestamp":"2017-09-27T12:13:05.018453866Z","message":"found free port 34201"}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","version":"13"},"level":"info","timestamp":"2017-09-27T12:13:05.01848986Z","message":"executing \"cd dist \u0026\u0026 node-v8.4.0-linux-x64/bin/node lib/functions\""}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","version":"13"},"level":"info","timestamp":"2017-09-27T12:13:05.024101628Z","message":"proxy (pid=12) started"}
{"fields":{"app":"HelpHero","plugin":"relay","region":"us-west-2","version":"13"},"level":"info","timestamp":"2017-09-27T12:13:05.024150704Z","message":"waiting for http://127.0.0.1:34201 to listen (timeout 15s)"}
```

Let me know if theres anything else that would be helpful

**Comment by tj** [comment]
@cloudkite interesting thanks! Looks like it is getting restarted, might have something to do with it

**Comment by tj** [comment]
@cloudkite also curious why it's restarting in the first place, indicates there's some problem with the Node app

**Comment by cloudkite** [comment]
@tj Thanks will try disabling the proxy, I also realised I'm not logging uncaught exceptions so I'll see if anything shows up there.

**Comment by cloudkite** [comment]
@tj after updating to v0.2.6 and turning off proxy retrying I no longer get the `Body length 0` error or any restarts in the logs however now I intermittently get

```
7:32:07am ERRO response duration: 0 id: a9da48dc-a60d-11e7-b0bd-69f819e1e570 method: POST path: /api/query size: 0 B stage: production status: 502 version: 17
```

Also nothing is being picked up by the `process.on('uncaughtException'` handler that I added.

I managed to replicate the 502 response once while invoking request through postman, however the exact same request succeeded the next time so I wasn't able to establish any kind of pattern.

**Comment by kaihendry** [comment]
I'm seeing up 0.2.6 return content-length zero too.

https://github.com/kaihendry/helloworld/tree/json

```
$ curl -sI https://hello.natalian.org/ | grep content-length
content-length: 0
x-amzn-remapped-content-length: 26
```

**Comment by tj** [comment]
@cloudkite hmm strange I tried

```
const http = require('http')
const fs = require('fs')
const { PORT = 3000 } = process.env

http.createServer((req, res) => {
  setTimeout(_ => {
    asdf()
    res.setHeader('Content-Type', 'application/json')
    fs.createReadStream('up.json').pipe(res)
  }, 50)
}).listen(PORT)

process.on('uncaughtException', function(err) {
  console.log('error %s', err)
  process.exit(1)
})
```

and it seems ok


```
  2:06:30pm INFO starting name: example-app type: server
  2:06:30pm INFO found free port 34533 version: 20
  2:06:30pm INFO executing "node app.js" version: 20
  2:06:30pm INFO proxy (pid=12) started version: 20
  2:06:30pm INFO waiting for http://127.0.0.1:34533 to listen (timeout 15s) version: 20
  2:06:31pm INFO error ReferenceError: asdf is not defined
  2:06:31pm ERRO response duration: 100 id: 93af79cc-a698-11e7-ab22-5fb74f471487 ip: 212.90.60.109 method: GET path: / size: 0 B stage: production status: 502 version: 20
```

**Comment by tj** [comment]
@kaihendry that's just API Gateway changing it for the HEAD request (`-I`), curl's `-D-` lets you output just the header for a GET:

```
$ curl -sD- `up url` | grep Content-Length
Content-Length: 12
x-amzn-Remapped-Content-Length: 12
```

**Comment by cloudkite** [comment]
@tj Do you think this is a logging problem or something else?

If logging is working correctly then that suggests there are no uncaught exceptions and something going wrong outside of the app in the shim or somewhere else?

Also no longer getting any mentions of restarts in the logs? Does that only happen when retrying is enabled?

Thanks for your help with this! if theres anything I can do on my end to help debug this please let me know.

**Comment by tj** [comment]
@cloudkite yep, Up should be only restarting if your process crashes. I guess the only other in-app problem could be if there's a process.exit(n) in there somewhere exiting silently, but that seems unlikely.

No problem! A few people have hit this, definitely need to dig into it. I should be back to work tomorrow so I'll do some more hunting and try to improve the debug-ability since I can't seem to reproduce this stuff easily

**Comment by tj** [comment]
- https://groups.google.com/forum/m/#!topic/golang-codereviews/pTrfhr7L15w

Go's reverse proxy code at the bottom there mentions that a zero-length response (only appropriate for a 204 response etc) results in a 502, so I suppose it could be your app after all, some kind of state issue. 

Either way it should be more obvious in Up that's for sure!

**Comment by cloudkite** [comment]
@tj I added the following log to my GraphQL API just before I stringify the body to see if I'm sending an empty or even falsy body at any point

```
    if (!body) {
      log.error('Empty body.', {
        id: req.headers['x-request-id'],
        status,
        headers,
        body,
      })
    }
```

using: `up logs -s 12h 'message = "Empty body."`
gives me a whole bunch of 204s so I know at least the logs for this are being entered.

using: `up logs -s 12h 'message = "Empty body." and status != 204'`
I get zero log entries however have still gotten a few 502s in same time period.
Unless there's something really weird going on then doesn't look like I'm ever returning an empty body.

btw the log search is really awesome 👍

**Comment by skabbes** [comment]
@tj you mentioned that Up only restarts your process if it crashes.  That is not true (in the current code), Up will restart your process once the proxy encounters a network error.  If you want that design to be true, I would be happy to work on it (it is a better design IMO).

**Comment by skabbes** [comment]
Ok, I have a fairly reliable pattern that I've found (and a hypothesis for why):

First, linking #313 the original issue logged and #334 since it was an initial try at fixing some of those problems.

**Forensic logs redacted with:**

```redact.js
const fs = require('fs');
const crypto = require('crypto');

const newlines = fs.readFileSync('./logs.json', 'utf8')
  .trim()
  .split('\n')
  .map(x => JSON.parse(x))
  .filter(l => l.fields.app !== true)
    .map((l) => {
    if (l.fields.path) {
      l.fields.path = crypto.createHash('sha256').update(l.fields.path, 'utf8').digest('base64').replace(/\=+$/g, '');
    }

    if (l.fields.query) {
      l.fields.query = crypto.createHash('sha256').update(l.fields.query, 'utf8').digest('base64').replace(/\=+$/g, '');
    }

    if (l.message && l.message.startsWith('executing')) {
      l.message = 'executing REDACTED';
    }

    return JSON.stringify(l);
  })
  .join('\n');

fs.writeFileSync('./redacted.txt', newlines);
```

[redacted.txt](https://github.com/apex/up/files/1356962/redacted.txt)

**Patterns:**
Connection reset by peer happens right before the http content length error.  And then the error happens a couple more times (presumably retries).  The connection reset always seems to happen after a _significant_ delay from the previous request (1 minute up to 5 minutes I think).

**Hypothesis:**
I think we are hitting the window where the container is _about_ to be discarded, but not yet (1-5 minutes of inactivity).  With that in mind, I think we either hitting up on bugs in the TCP stacks / http stacks of the app code, where they cannot really operate as they should after being suspended for so long.  Possibly the client app code (hapi.js for @berzniz and vanilla go for me) or amazonlinux tcp stack code is sending an RST over the connection (which after so much inactivity would probably be correct). The normal FIN-ACK sequence will not work since both sides of the TCP connection are suspended.

I think that because up-proxy is using a reverse proxy with agressive connection reuse, it _tries_ to send a request to the application code and immediately receives an RST.  Then _I believe_ this is exposing a bug with the retry logic (causing the http contentlength errors).

Overall, I believe the proxy has some version of [this bug](https://medium.com/@xoen/golang-read-from-an-io-readwriter-without-loosing-its-content-2c6911805361) reading from a closed body gives nothing.

**Comment by cloudkite** [comment]
@tj @skabbes 

No sure if this helps, but another thing I tried todo is logging every request inside the server callback

```
http.createServer((req, res) => {
  log.info('hello', { id: req.headers['x-request-id'] })
  handler(req, res)
}).listen(port)
```

for the requests ids that respond with 502 i never see the "hello" log, every other request gets a "hello" log entry

**Comment by tj** [comment]
@cloudkite I noticed in the stdlib that the code path for this content-length/body error is used for both the request and response, they re-use the same code, so it definitely could be likely that it never reaches your app, like @skabbes mentions about the immediate reset. 

I swear net/http's Transport does some retrying internally on connection resets, I remember reading that it only bails on a partial write. So maybe we're hitting that case, where part of the request body is written, and Transport thinks that it can no longer safely re-transmit. 

@skabbes thanks again for digging into it! I've been sick the past week so I haven't had much time. Sounds like from @cloudkite's testing that it's the request and lines up with what you're suggesting.

**Comment by tj** [comment]
need to do a more thorough read, but I think this is the spot https://github.com/golang/go/blob/ec40c7e189182918480099bd73b5c0c73018736c/src/net/http/transport.go?ts=2#L469-L472

**Comment by skabbes** [comment]
If you have some code you want to throw into a prod environment that _will_ repo this, let me know - I can run a custom version of up.

**Comment by tj** [comment]
Maybe we should try disabling net.Dialer's keepalive, I suppose if the container is "frozen" that may not be functioning properly. Bit of a wild guess but maybe worth a shot

**Comment by skabbes** [comment]
AHH, @tj you know what it might be??

I bet for POSTs, we are re-reading the _request_ body twice.

1. Request comes in, with a body. [create a faux http.request](https://github.com/apex/up/blob/master/internal/proxy/request.go#L28-L41)
1. We send it to the reverse proxy (and it consumes the request body)
1. We need to [retry](https://github.com/apex/up/blob/master/http/relay/relay.go#L144) the request, but cannot read the body anymore - Error.

No matter what happens after 3, we can never re-read the body of the request, so we will never complete it.

**Comment by cloudkite** [comment]
@skabbes Does the retry workflow you described still happen if you have `"proxy": { "retry": false }`? as I have retries disabled and am still getting this

**Comment by skabbes** [comment]
@cloudkite no, I haven't yet - but I can.  Do you have any logs that I can take a look at?  Feel free to use the redact script I have above.

**Comment by tj** [comment]
Hmm yea seems like @cloudkite hits it without retries though that's the extra strange part haha. We could try that `req.GetBody` thing to recycle the body maybe, but it doesn't explain sans-retry errors

**Comment by berzniz** [comment]
I’m also experiencing it without retries. However, it only happens for one PUT endpoint out of about 30 endpoints. Tried logging it, but I’m not sure I’m seeing the request reaching the node server

**Comment by skabbes** [comment]
I don't have time today to debug, but I'm going to at least deploy with "retry: false" today and get logs for tomorrow.

**Comment by berzniz** [comment]
More observations I see:

Since API Gateway has a timeout of 30 seconds, I also set the Lambda's timeout to 30 seconds. Some requests take more than 30 seconds (I have no idea why, they are not doing much). When these requests finish and respond, the connection is reset.

You can see here in the log that `endpoint1` gets a `502` at the exact moment where `endpoint2` responds (`endpoint2` request started long before the request for `endpoint1`).

```
INFO  2017-10-06T09:36:46.223Z (deb46ae8-aa79-11e7-9499-dfe0ba962dc3) Verifying Usage Plan for request: deb46ae8-aa79-11e7-9499-dfe0ba962dc3. API Key:  API Stage: b98tge5a8h/production 
INFO  2017-10-06T09:36:46.223Z (deb46ae8-aa79-11e7-9499-dfe0ba962dc3) API Key  authorized because method 'ANY /{proxy+}' does not require API Key. Request will not contribute to throttle or quota limits 
INFO  2017-10-06T09:36:46.223Z (deb46ae8-aa79-11e7-9499-dfe0ba962dc3) Usage Plan check succeeded for API Key  and API Stage b98tge5a8h/production 
INFO  2017-10-06T09:36:46.223Z (deb46ae8-aa79-11e7-9499-dfe0ba962dc3) Starting execution for request: deb46ae8-aa79-11e7-9499-dfe0ba962dc3 
INFO  2017-10-06T09:36:46.223Z (deb46ae8-aa79-11e7-9499-dfe0ba962dc3) HTTP Method: PUT, Resource Path: /api/endpoint1
INFO  2017-10-06T09:36:46.236Z START RequestId: deb4b96e-aa79-11e7-af84-bd8064c08cec Version: 74 
INFO  2017-10-06T09:36:46.249Z 171006/093410.371, [response,api] http://ip-10-20-36-172:33499: post /api/endpoint2 {} 200 (32368ms) 
INFO  2017-10-06T09:36:46.251Z 2017/10/06 09:36:46 http: proxy error: EOF 
ERROR 2017-10-06T09:36:46.251Z PUT /apiendpoint1 (502) response 
INFO  2017-10-06T09:36:46.252Z END RequestId: deb4b96e-aa79-11e7-af84-bd8064c08cec 
INFO  2017-10-06T09:36:46.252Z REPORT RequestId: deb4b96e-aa79-11e7-af84-bd8064c08cec  Duration: 11.85 ms  Billed Duration: 100 ms   Memory Size: 768 MB Max Memory Used: 176 MB 
INFO  2017-10-06T09:36:46.256Z (deb46ae8-aa79-11e7-9499-dfe0ba962dc3) Successfully completed execution 
INFO  2017-10-06T09:36:46.256Z (deb46ae8-aa79-11e7-9499-dfe0ba962dc3) Method completed with status: 502 
```

**Comment by skabbes** [comment]
Ah, this is all starting to make sense. At least the layers of timeouts,
process suspension etc are starting to become more clear

**Comment by berzniz** [comment]
I’m looking at the `_proxy.js` shim file (https://github.com/apex/up/blob/master/internal/shim/index.js) and wondering: `var callback` seems to always remember just the last invocation’s `cb` function. If two requests comes one after the other, the first one could potentially responds to the `cb` of the second and never respond to its own `cb`.

Could this lead to a horrible bug or am I missing something?

**Comment by skabbes** [comment]
@berzniz I think you're right, it would explain a lot going on here.  I think the design of that shim is that lambda will only call your handler once concurrently, but sounds like it wasn't 100% thought through.

I'm going to prototype today up examples with various timeout characteristics, and try to find a minimal repro.

**Comment by berzniz** [comment]
I've setup this endpoint here:
https://k372i6uy51.execute-api.us-west-2.amazonaws.com/development/:timeout

Where `timeout` is in milliseconds, for example:
https://k372i6uy51.execute-api.us-west-2.amazonaws.com/development/5000 -> will wait 5 seconds and responds with:
```
{"timeout":5000,"start":"2017-10-08T20:36:07.120Z","end":"2017-10-08T20:36:12.125Z"}
```
I tried it out but could not reproduce. I'm keeping the example up, here's the code:
```
const express = require('express')
const app = express()
const { PORT = 3000 } = process.env

app.get('/:time', (req, res) => {
  let { time } = req.params
  const timeout = parseInt(time, 10) || 0
  const start = new Date()
  setTimeout(() => {
    const end = new Date()
    res.send({ timeout, start, end })
  }, timeout)
})

app.listen(PORT, () => {
  console.log(`Example app listening on port ${PORT}!`)
})
```

**Comment by skabbes** [comment]
Hey Everyone, great news I have been able to reproduce contentlength 0 error with [this code](https://github.com/skabbes/up-examples/tree/bug/502_proxy/oss/node-proxy-fail) and a slight modification to up to allow POSTs to be retried.  The error seems to come when a request's body is re-read during a retry (Important to note, that even if your code is perfect, retries will still be necessary since suspending the lambda process suspending will really screw with your idle tcp connection on the reverse proxy).

This code fixes it: https://github.com/apex/up/pull/363

Basically go's default http transport allows for re-reading the request body with an optional function `GetBody`.  We simply needed to implement that.

@tj it would be great to get a release of this ASAP, since I spoke with @berzniz tonight and sounds like he's about read to ditch Up in production because of this bug.

I am running a custom build, so it is less urgent for me.

**Comment by cloudkite** [comment]
@skabbes awesome work! looking forward to giving it a test

**Comment by tj** [comment]
Doing some QA and I'll have those changes in a release soon!

**Comment by tj** [comment]
`up upgrade` for 0.2.7 with @skabbes' fixes, definitely let us know if that does that trick, note you'll have to change `.lambda.timeout` to `.proxy.timeout` (lambda stuff is implied now)

**Comment by cloudkite** [comment]
@tj I upgraded to 0.2.8 and moved the timeout to the proxy section of config and re-enabled retry. however I still get 502's :(

```
12:33:27pm INFO response duration: 19 id: 403798f2-ad4a-11e7-a949-abe57b013739 method: POST path: /api/query query: trace=nav-startVisit size: 1.3 kB stage: production status: 200 version: 37
  12:33:33pm WARN temporary error: read tcp 127.0.0.1:51190->127.0.0.1:34961: read: connection reset by peer version: 37
  12:33:33pm ERRO network error: http: ContentLength=1131 with Body length 0 version: 37
  12:33:33pm WARN restarting version: 37
  12:33:33pm INFO found free port 40329 version: 37
  12:33:33pm INFO executing "cd dist && node-v8.6.0-linux-x64/bin/node lib/functions" version: 37
  12:33:33pm INFO proxy (pid=28) started version: 37
  12:33:33pm INFO waiting for http://127.0.0.1:40329 to listen (timeout 15s) version: 37
  12:33:34pm WARN restarted restarts: 1 version: 37
  12:33:34pm ERRO response duration: 976 id: 43c63e6d-ad4a-11e7-8bb5-7bf6da8d0b3e method: POST path: /api/query query: trace=nav-startVisit size: 0 B stage: production status: 502 version: 37
  12:33:42pm INFO response duration: 45 id: 4946eebc-ad4a-11e7-8786-4b664fc4e53d method: GET path: /embed/DLePgo8rWPR size: 55 B stage: production status: 302 version: 37
```

**Comment by cloudkite** [comment]
might be related to https://github.com/golang/go/issues/22158

**Comment by skabbes** [comment]
I had a similar piece of software to up that simply created a new http.client for every request.  it isn't ideal perf-wise, but it worked reliably.  Would anyone be up for that instead?

The go issue seems to be exactly the same behavior that is being exhibited here FWIW.

**Comment by cloudkite** [comment]
My understanding is that lambda does not processes requests in parallel instead it starts other instances to handle concurrency? in which case would it even make a difference perf-wise?

**Comment by skabbes** [comment]
I suppose very little, the overhead of establishing a new connection. The
initial TCP handshake. My vote it to try that next.

**Comment by skabbes** [comment]
Ok, I have another attempt at this:

https://github.com/apex/up/pull/365

The change is rather trivial, but the behaviors described and the resolutions match almost exactly.  I've deployed it to prod, fingers crossed.

**Comment by tj** [comment]
Hmmm it does seem to have a noticeable impact on latency, seeing more timeouts now, at least with some quick non-thorough testing

**Comment by tj** [comment]
Weird deployed a few more times and I think it's ok, false alarm hopefully :D. Will do some more thorough testing there to make sure we didn't compromise latency, not being broken is more important of course haha

**Comment by skabbes** [comment]
Oh man, I'd be incredibly skeptical of lambda if their loopback network interface was that slow.

I think overall, this is likely a win for 99p latency, since it prevents the http bug (which incurs the additional latency of restarting a process), and faster responses lead to more container reuse.  And I betcha booting up a new container is a lot more latency than a TCP handshake :)

**Comment by tj** [comment]
oh yea definitely, it was just odd since I was getting zero timeouts with previous versions, but seems like just that one deploy was messed for whatever reason, just Lambda acting up or something

**Comment by skabbes** [comment]
Ah, cool. You scared me a little bit there.

**Comment by cloudkite** [comment]
Can confirm that v0.2.9 has resolved it for me 👍 👯 
Thanks @tj, @skabbes!

**Comment by komuW** [comment]
From someone who has been watching this from the sidelines, I would like to
say i appreciate all the effort from everyone involved 👌

On Oct 11, 2017 7:25 AM, "Jonas Budelmann" <notifications@github.com> wrote:

Can confirm that v0.2.9 has resolved it for me 👍 👯
Thanks @tj <https://github.com/tj>, @skabbes <https://github.com/skabbes>!

—
You are receiving this because you are subscribed to this thread.
Reply to this email directly, view it on GitHub
<https://github.com/apex/up/issues/276#issuecomment-335678550>, or mute the
thread

**Comment by tj** [comment]
Big thanks to everyone from me too, and early adopters of course! Super helpful to have all the feedback

**Comment by tj** [comment]
Sounds like we're good to close, thanks everyone!
